### PR TITLE
refactor: avoid requiring time predicate in Segment

### PR DESCRIPTION
**Pull request by e-dard**
Things would be a bit simpler inside a `Segment` if there wasn't a requirement to have predicates on the `time` column. Enforcing the requirement to range by time can still be achieved, it would just be done at the `table` level rather than the segment level.

For segments that are completely covered by a time-range predicate(s) the `table` can then just remove those predicates before pushing down the rest of the predicates to the relevant segment.

### CODE REVIEW

File: segment_store/src/segment.rs [hunk 1]
```diff
@@ -293,11 +298,34 @@ impl Segment {
...
+            // All rows matched all predicates because any predictates not matching
+            // any rows would have resulted in an early return.
             return RowIDsOption::All(result_row_ids);
         }
         RowIDsOption::Some(result_row_ids)
     }
+
+    // An optimised function for applying two comparison predicates to a time
+    // column at once.
+    fn row_ids_from_predicates_with_time_range(
```
Comment (e-dard) [inline segment_store/src/segment.rs:310]: This is called if there are two predicates on the time column. There is a dedicates method for applying both of these predicates together on the time column.

File: segment_store/src/segment.rs [hunk 2]
```diff
@@ -208,67 +208,68 @@ impl Segment {
         }
     }
 
-    // Determines the set of row ids that satisfy the time range and all of the
-    // optional predicates.
-    //
-    // TODO(edd): right now `time_range` is special cased so we can use the
```
Comment (e-dard) [inline segment_store/src/segment.rs:214]: Bye bye

File: segment_store/src/segment.rs [hunk 3]
```diff
@@ -208,67 +208,68 @@ impl Segment {
...
-        // TODO(edd): potentially pass this in so we can re-use it once we
-        // have materialised any results.
-        let mut result_row_ids = RowIDs::new_bitmap();
-
-        match time_row_ids {
-            // No matching rows based on time range - return buffer
-            RowIDsOption::None(_) => return time_row_ids,
+            .filter(|(col, _)| *col == TIME_COLUMN_NAME)
+            .count()
+            == 2
```
Comment (e-dard) [inline segment_store/src/segment.rs:235]: Check if we have two predicates on the time column, i.e., a time range.

File: segment_store/src/segment.rs (hunk 2 above)
Comment (alamb) [reply segment_store/src/segment.rs:214]: It is almost like this comment was out of date - as in the time predicate wasn't special cased in the function signature

File: segment_store/src/segment.rs (hunk 3 above)
Comment (alamb) [reply segment_store/src/segment.rs:235]:
```suggestion
            // Check if we have two predicates on the time column, i.e., a time range.
            == 2
```

File: segment_store/src/segment.rs [hunk 4]
```diff
@@ -470,10 +494,76 @@ mod test {
...
 
-        arr.extend(column_predicates);
+        arr.extend(others);
         arr
     }
 
+    #[test]
+    fn row_ids_from_predicates() {
+        let mut columns = BTreeMap::new();
+        let tc = ColumnType::Time(Column::from(&[100_i64, 200, 500, 600, 300, 300][..]));
```
Comment (alamb) [inline segment_store/src/segment.rs:504]: This `[..]` threw me -- but then I tried to remove it and I got some strange error about there not being a conversion from array to slice....  

```
   --> segment_store/src/segment.rs:504:35
    |
504 |         let tc = ColumnType::Time(Column::from(&[100i16, 200, 500, 600, 300, 300]));
    |                                   ^^^^^^^^^^^^ the trait `From<&[i16; 6]>` is not implemented for `column::Column`
    |
    = help: the following implementations were found:
              <column::Column as From<&[&str]>>
              <column::Column as From<&[Option<&str>]>>
              <column::Column as From<&[f64]>>
              <column::Column as From<&[i16]>>
            and 9 others
```

File: segment_store/src/segment.rs (hunk 3 above)
Comment (alamb) [reply segment_store/src/segment.rs:235]: I also wonder would happen if there are three predicates on the timestamp? `ts > 5 && ts < 10 && ts != 7` -- probably not really worth worrying about, but eventually finding the range timestamps might need to be a bit more clever

**Review by alamb — APPROVED** [review]
I reviewed this and it looks good to me. 👍

**Pull request merged**